src/board/arduino/mega2560/parallel.c

```diff
@@ -228,6 +249,55 @@ int parallel_transaction(struct Parallel * port, uint8_t * data, int length, boo
 #define parallel_read(P, D, L) parallel_transaction(P, D, L, true, false)
 #define parallel_write(P, D, L) parallel_transaction(P, D, L, false, false)
 
+// host write -> peripheral read
+// host read -> peripheral write
+int parallel_peripheral_transaction(struct Parallel * port, uint8_t * data, int length, bool read, bool * addr) {
+    if (!read) {
+        // Set data lines as outputs to write to host
+        parallel_data_forward(port);
+    }
+
+    int i;
+    uint8_t byte = 0;
+    for (i = 0; i < length; i++) {
+        // Wait for host to indicate it's written data
+        while (gpio_get(port->data_n) && gpio_get(port->addr_n)) {}
+
+        *addr = !gpio_get(port->addr_n);
+
+        if (read) {
```

Comment:
Read/write direction is controlled by the host with the write_n line. If this line is low, then the host is writing to the peripheral.